### ISSUE TITLE
Add support for scheduled batch ingestion

### ISSUE THREAD

**Issue by woop** [feast-dev/feast#1180]
The current Job Service only allows for batch ingestion based on an [external trigger.](https://github.com/feast-dev/feast/blob/master/protos/feast/core/JobService.proto#L30) This card proposes the expansion of feature table definitions in order to allow for scheduled batch ingestion.

Before
```
driver_ft = FeatureTable(
    name="driver_trips",
    entities=["driver_id"],
    features=[
      Feature("average_daily_rides", ValueType.FLOAT),
      Feature("rating", ValueType.FLOAT)
    ],
    batch_source=BigQuerySource(
        table_ref="gcp_project:bq_dataset.bq_table",
    )
)
```

After
```
driver_ft = FeatureTable(
    name="driver_trips",
    entities=["driver_id"],
    features=[
      Feature("average_daily_rides", ValueType.FLOAT),
      Feature("rating", ValueType.FLOAT)
    ],
    batch_source=BigQuerySource(
        table_ref="gcp_project:bq_dataset.bq_table",
    ),
   batch_schedule="1d", 
   online=True,
   feature_start_time=datetime(1, 1, 20)
)
```

- `batch_schedule`: Interval at which jobs will be launched. This schedule will be used along with the `feature_start_time` to determine how many jobs to launch when the job service starts up. After that, new jobs will only be launched as time passes the scheduled interval.
- `feature_start_time`: The date from which scheduled jobs should start.
- `online`: Whether ingested data should be pushed to the online store. Since we don't have an offline store right now, this key must always be set to `True`. Setting it to `False` should result in an exception.

As part of this implementation we should try to not maintain state within the job service. Ideally we would be able to track all jobs that have been launched purely by looking at historic jobs on the Jobs API (EMR, Dataproc)

**Comment by oavdeev-tt** [comment]
A couple things we should think through before starting on this:

### State
I think it would be tough to not introduce state for this one. At least for EMR I believe historic jobs are only visible for some limited, not well defined period of time.

### Scope
Another thing is I think it is hard to rely on this functionality in practice without having debug tooling such as
* a way see which jobs have failed
* some way to retry them on-demand (if they failed due to misconfiguration on users end)
* configure automatic retries

So in the end this starts to look like a reimplementation of Airflow. 

Some suggestions to avoid the scope creep that come to mind:
1. we could just not do this and leave it to the user
2. and/or provide Airflow operator for Feast
3. and/or implement this but consider it a "toy" scheduler that is explicitly not intended to be used in production. Answer every future feature request with "you should really just use Airflow".

**Comment by woop** [comment]
State: If we don't have long running state for EMR, then it seems like its inevitable that we do need to have some state tracking here. I also dont think the need for state goes away by not picking up scheduling. 

The user experience would be pretty poor if we asked them to track all failed/successful jobs on their side. I am still a bit 50/50 on where to draw the line here, but it seems like at least being able to tell users which jobs have failed and succeeded should be the minimum. Being able to group jobs by their unique interval/start date seems useful as well, but I am not sure if that should be on our side. 

Anyway, the use case I am thinking of is having an `ensure`-like command that applies to both batch and streaming. I like the idea of an admin being able to do something like
```
feast jobs ensure --all
```
from a single scheduler/cron or simply their local machine. It would stop old/outdated jobs and spin up missing jobs, both for streaming and batch.

We could also add a remote repository or backend for state
```
feast jobs ensure --backend gs://feast-state/ --repo https://github.com/acme/feast-repo --all 
```
Now that I think about it, `ensure` is very similar to Tecton `apply`.

@oavdeev I'm open to all your suggestions. I also feel like limiting the scope here is important since we dont want to reinvent Airflow. Which of the following do you think we should cut/keep the scope of?

1. Scheduling built into Feast
1. Identifying batch jobs based on their schedule interval. (user would need to set the `batch_schedule` and they would be able to trigger a specific historical job based on its start_date or some timestamp, but not select arbitrary time ranges for ingestion). 
1. Tracking the state of batch jobs that have historically run, but ideally using the local filesystem or a bucket, and not Postgres.
1. Adding a batch/stream `ensure`/`apply` method

I am leaning toward keeping (2), (3), and (4). In theory we could remove (2) and always just ingest the missing data since the previous run. However, not tracking the schedule interval would make aggregations impossible.

**Comment by oavdeev** [comment]
So we'd keep track of what date ranges were ingested but it is up to you to actually trigger the jobs? (by calling that "ensure") That seems reasonable.

I'd point out that if you're using Airflow, you'd likely _not_ want Feast to keep track of that because it is already tracked in Airflow state. I'm not sure how to nicely address that yet.

**Comment by woop** [comment]
> I'd point out that if you're using Airflow, you'd likely not want Feast to keep track of that because it is already tracked in Airflow state. I'm not sure how to nicely address that yet.

I dont think there is any way to have an `ensure` without maintaining state, unless Feast calls Airflow which I am not sure how difficult it would be. It also adds more required infrastructure which is something I want to move away from for a minimal setup.

How about we do the following

We skip (1) and (2) above, meaning Feast doesn't schedule jobs and doesnt track intervals. It only does (3) and (4). If you an `ensure` method we run a batch job between `feature_start_time` and `now()`, and we track the final event timestamp that was ingested (the `now()` essentially). Rerunning `ensure` then ingests the difference between the old end date and the new `now()`. So the minimal state we track is basically a timestamp for each feature table, much like streaming ingestions have consumer offsets. Although I think its still nice to be able to track state that allows commands like `feast jobs list` to print out meaningful output.

So maybe teams running Airflow could
- Set up a single trigger that just runs `ensure`. 
- or, Set up a `feast job start --feature-table my_feature_table --start_date {{ airflow start date }} --end_date {{ start + schedule_interval }}` for each feature table, and then specify the start and end dates based on Airflow schedules.

I think these teams can just ignore the job state being tracked within Feast if they are using Airflow, or are you concerned about an inconsistency arriving there @oavdeev?

**Comment by oavdeev** [comment]
> I think these teams can just ignore the job state being tracked within Feast if they are using Airflow, or are you concerned about an inconsistency arriving there @oavdeev?

Exactly, I'm a bit worried if Airflow state and Feast state somehow desync, it would be hard to notice and debug.

But yes we could keep the current `start_offline_to_online_ingestion` (low level method, no state tracking, to be used with Airflow) but then also add a more high-level `ensure` (name TBD) that would keep track of state.

**Comment by woop** [comment]
> > I think these teams can just ignore the job state being tracked within Feast if they are using Airflow, or are you concerned about an inconsistency arriving there @oavdeev?
> 
> Exactly, I'm a bit worried if Airflow state and Feast state somehow desync, it would be hard to notice and debug.
> 
> But yes we could keep the current `start_offline_to_online_ingestion` (low level method, no state tracking, to be used with Airflow) but then also add a more high-level `ensure` (name TBD) that would keep track of state.

Just quickly want to clarify what you mean by "no state tracking". We are still tracking these jobs on EMR and exposing references and logs (essentially state) to users. It's just that we don't maintain that separately as well in a bucket or database, so that if the EMR cluster is destroyed then the job info is lost.

Is that what you mean?

**Comment by woop** [comment]
We are pushing out this card for the time being.
